hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/placement/algorithms/SCMContainerPlacementRackScatter.java

```diff
@@ -0,0 +1,343 @@
+   * meets all the requirements, there is fallback chosen process depending on
+   * whether fallback is allowed when this class is instantiated.
+   *
+   *
+   * @param excludedNodes - list of the datanodes to excluded. Can be null.
+   * @param affinityNode - the chosen nodes should be on the same rack as
+   *                    affinityNode. Can be null.
+   * @param dataSizeRequired - size required for the container.
+   * @param metadataSizeRequired - size required for Ratis metadata.
+   * @return List of chosen datanodes.
+   * @throws SCMException  SCMException
+   */
+  private Node chooseNode(List<Node> excludedNodes, Node affinityNode,
+      long metadataSizeRequired, long dataSizeRequired) throws SCMException {
+    int ancestorGen = RACK_LEVEL;
+    int maxRetry = MAX_RETRY;
+    List<String> excludedNodesForCapacity = null;
+    while(true) {
+      metrics.incrDatanodeChooseAttemptCount();
+      Node node = networkTopology.chooseRandom(NetConstants.ROOT,
+          excludedNodesForCapacity, excludedNodes, affinityNode, ancestorGen);
+      if (node == null) {
+        // cannot find the node which meets all constrains
+        LOG.warn("Failed to find the datanode for container. excludedNodes:" +
+            (excludedNodes == null ? "" : excludedNodes.toString()) +
+            ", affinityNode:" +
+            (affinityNode == null ? "" : affinityNode.getNetworkFullPath()));
+        return null;
+      }
+
+      DatanodeDetails datanodeDetails = (DatanodeDetails)node;
+      if (isValidNode(datanodeDetails, metadataSizeRequired,
+          dataSizeRequired)) {
+        return node;
+      }
+
+      maxRetry--;
+      if (maxRetry == 0) {
+        // avoid the infinite loop
+        String errMsg = "No satisfied datanode to meet the space constrains. "
+            + "metadatadata size required: " + metadataSizeRequired +
+            " data size required: " + dataSizeRequired;
+        LOG.info(errMsg);
+        return null;
+      }
+      if (excludedNodesForCapacity == null) {
+        excludedNodesForCapacity = new ArrayList<>();
+      }
+      excludedNodesForCapacity.add(node.getNetworkFullPath());
+    }
+  }
+
+  /**
+   * This default implementation handles rack aware policies and non rack
+   * aware policies. If a future placement policy needs to check more than racks
+   * to validate the policy (eg node groups, HDFS like upgrade domain) this
+   * method should be overridden in the sub class.
+   * This method requires that subclasses which implement rack aware policies
+   * override the default method getRequiredRackCount and getNetworkTopology.
+   * @param dns List of datanodes holding a replica of the container
+   * @param replicas The expected number of replicas
+   * @return ContainerPlacementStatus indicating if the placement policy is
+   *         met or not. Not this only considers the rack count and not the
+   *         number of replicas.
+   */
+  @Override
```

Comment:
This method is almost a duplicate of the one in SCMCommonPlacementPolicy. My intention of that implementation was that the method getRequiredRackCount() should be overridden in the sub-classes, and hopefully then this method does not need to be overridden. However getRequiredRackCount() does not work for EC, as we need to pass the number of replicas to it.

I think it would be better, if we changed `SCMCommonPlacementPolicy.getRequiredRackCount()` to accept a parameter `int numReplicas`, and then the sub-classes can ignore or use that. That means in this class, the implementation would be:

```
@Override
protected int getRequiredRackCount(int expectedReplicas) {
    return expectedReplicas;
  }
```

Do you think that makes sense?

Reply:
One other suggestion here, which I should have mentioned before. It would be a good idea to make this change on master branch first, and then we can merge master to the EC branch. That way, we will avoid potential conflicts in the future if someone makes some further changes related to this method. Same for the `isValidNode()` suggestion. We could make these two changes in a single PR against master. We have this Jira HDDS-5351 where we have done similar things already when developing EC.

Reply:
Sure, I will raise another ticket to add the changes to master.